apps/prairielearn/src/lib/code-caller/code-caller-container.js

```diff
@@ -333,34 +355,25 @@ export class CodeCallerContainer {
     this.debug('exit done()');
   }
 
+  /** @private */
   async ensureChild() {
     this.debug('enter ensureChild()');
     this._checkState();
 
-    // Since container creation is async, it's possible that ensureChild()
-    // could be called again while it's already executing. For instance, we
-    // could be inside a call that was made to warm up this worker, but then
-    // we might have call() invoked, which will also call ensureChild(). We
-    // need to ensure that we're only ever inside this function once at a time,
-    // so we'll use a "mutex" (even though we're in a single-threaded environment).
-    await this.ensureChildMutex.runExclusive(async () => {
-      if (this.container) {
-        this.debug('exit ensureChild() - existing container');
-        return;
-      }
-      this.hostDirectory = await tmp.dir({ unsafeCleanup: true, prefix: MOUNT_DIRECTORY_PREFIX });
-      await ensureImage();
-      await this._createAndAttachContainer(this.hostDirectory.path);
+    if (this.container) {
+      this.debug('exit ensureChild() - existing container');
+      return;
+    }
```

Comment:
I don't quite understand why the old code needed the mutex. Can you help me understand why it used to be needed but isn't anymore?

Reply:
In the version immediately prior to this PR, it actually wasn't strictly needed either. IIRC the mutex was added before the pool was responsible for fully warming up workers. In the past, `ensureChild()` was a fire-and-forget operation where the pool would call it but then immediately return the new caller, so it's possible that someone would then call `call(...)` while the worker was still warming up, which would then in turn call `ensureChild()` again, so you could end up with two concurrent calls.

Now, `ensureChild()` is called only in the `create()` function, and only a single time, so we don't have to worry about this.

Does that clear things up?